src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -4163,6 +4356,11 @@ Pipe MergeTreeData::alterPartition(
     const PartitionCommands & commands,
     ContextPtr query_context)
 {
+    /// Wait for loading of outdated parts
+    /// because partition commands (DROP, MOVE, etc.)
+    /// must be applied to all parts on disk.
+    waitForOutdatedPartsToBeLoaded();
```

Comment:
`alterPartition()` will not be called from `execureDropRange`/`executeReplaceRange`, need to add this call there as well

Reply:
It's easy to forget to call it, so let's add `assert(outdated_data_parts_loading_finished)` to `removePartsInRangeFromWorkingSetAndGetPartsToRemoveFromZooKeeper` (it's my favorite method name, btw)